pkg/k8s/resource/resource.go

```diff
@@ -76,28 +74,23 @@ type Resource[T k8sRuntime.Object] interface {
 //		// Handle error ...
 //	}
 //
-// See also pkg/k8s/resource/example/main.go.
-func NewResourceConstructor[T k8sRuntime.Object](lw func(c k8sClient.Clientset) cache.ListerWatcher) func(lc hive.Lifecycle, c k8sClient.Clientset) Resource[T] {
-	return NewResourceConstructorWithRateLimiter[T](workqueue.DefaultControllerRateLimiter(), lw)
-}
 
-// NewResourceConstructorWithRateLimiter is the same as NewResourceConstructor, but with a custom rate limiter.
-func NewResourceConstructorWithRateLimiter[T k8sRuntime.Object](rateLimiter workqueue.RateLimiter, lw func(c k8sClient.Clientset) cache.ListerWatcher) func(lc hive.Lifecycle, c k8sClient.Clientset) Resource[T] {
-	return func(lc hive.Lifecycle, c k8sClient.Clientset) Resource[T] {
-		if !c.IsEnabled() {
-			return nil
-		}
-		newLW := func() cache.ListerWatcher { return lw(c) }
-		res := newResource[T](newLW, rateLimiter)
-		lc.Append(hive.Hook{OnStart: res.start, OnStop: res.stop})
-		return res
+// See also pkg/k8s/resource/example/main.go for a runnable example.
+func New[T k8sRuntime.Object](lc hive.Lifecycle, lw cache.ListerWatcher, opts ...Option) Resource[T] {
+	r := &resource[T]{
+		queues: make(map[uint64]*keyQueue),
+		opts:   defaultOptions(),
+		lw:     lw,
 	}
+	for _, applyOpt := range opts {
+		applyOpt(&r.opts)
+	}
+	r.ctx, r.cancel = context.WithCancel(context.Background())
```

Comment:
What is this context used for? Shouldn't it be provided by `OnStart()`?

Reply:
No the ctx in start is only valid for the duration of the start itself. It is used to cancel start if things take too long to setup. E.g. say a database module can block on start to connect to a database and if it takes too long the start can be aborted by cancelling the start context.

This context here is for the whole duration of the resource and is cancelled when stopping.

I’ve for the whole time I’ve worked with fx/dig been wondering whether we should just remove the context from the start as it is confusing. The alternative is to not have it and then just ”fatal” if start gets stuck. We wouldn’t have the opportunity to then clean up things, but we haven’t needed that thus far anyway…

Reply:
Oof, that's very, very confusing. I would assume that `Start(ctx)` receives the "app-level" context, not the "startup-level" context. I understand the utility of having separate "hook" contexts vs. "app" contexts, especially for shutdown.

I suppose you could catch applications that misbehave by immediately cancelling the Start context after all hooks are done. But perhaps better is just to wire-in an app-level or hive-level context?

Honestly, a stop context makes some sense to me, but not a start context.

Reply:
Thought about this a bit. I really don't like the idea of removing the start context as there's plenty of places where we do blocking operations from the start hook: for example the k8s-client blocks until it actually can connect to the api-server, local node info is initialized at start from api-server and so on. We really should handle these sort of timeouts gracefully and provide clean reasons in the logs what went wrong, e.g. "this start hook timed out" rather than Fatal with non-specific error message + dump stack trace, and to give the possibility to gracefully stop modules started before it.

So I've now added `hive.HookContext` as a new type of context.Context and documented its purpose. Hope that clears up the confusion in the future.